src/rust/src/x509/verify.rs

```diff
@@ -180,6 +215,107 @@ self_cell::self_cell!(
     }
 );
 
+#[pyo3::pyclass(
+    frozen,
+    name = "VerifiedClient",
+    module = "cryptography.hazmat.bindings._rust.x509"
+)]
+struct PyVerifiedClient {
+    #[pyo3(get)]
+    subject: pyo3::Py<pyo3::PyAny>,
+    #[pyo3(get)]
+    chain: pyo3::Py<pyo3::types::PyList>,
+}
+
+#[pyo3::pyclass(
+    frozen,
+    name = "ClientVerifier",
+    module = "cryptography.hazmat.bindings._rust.x509"
+)]
+struct PyClientVerifier {
+    policy: OwnedPolicy,
+    #[pyo3(get)]
+    store: pyo3::Py<PyStore>,
+}
+
+impl PyClientVerifier {
+    fn as_policy(&self) -> &Policy<'_, PyCryptoOps> {
+        &self.policy.borrow_dependent().0
+    }
+}
+
+#[pyo3::pymethods]
+impl PyClientVerifier {
+    #[getter]
+    fn validation_time<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        datetime_to_py(py, &self.as_policy().validation_time)
+    }
+
+    #[getter]
+    fn max_chain_depth(&self) -> u8 {
+        self.as_policy().max_chain_depth
+    }
+
+    fn verify(
+        &self,
+        py: pyo3::Python<'_>,
+        leaf: pyo3::Py<PyCertificate>,
+        intermediates: Vec<pyo3::Py<PyCertificate>>,
+    ) -> CryptographyResult<PyVerifiedClient> {
+        let policy = self.as_policy();
+        let store = self.store.get();
+
+        let chain = cryptography_x509_verification::verify(
+            &VerificationCertificate::new(
+                leaf.get().raw.borrow_dependent().clone(),
+                leaf.clone_ref(py),
+            ),
+            intermediates.iter().map(|i| {
+                VerificationCertificate::new(
+                    i.get().raw.borrow_dependent().clone(),
+                    i.clone_ref(py),
+                )
+            }),
+            policy,
+            store.raw.borrow_dependent(),
+        )
+        .map_err(|e| VerificationError::new_err(format!("validation failed: {e:?}")))?;
+
+        let py_chain = pyo3::types::PyList::empty(py);
+        for c in &chain {
+            py_chain.append(c.extra())?;
+        }
+
+        // NOTE: These `unwrap()` cannot fail, since the underlying policy
+        // enforces the presence of a SAN and the well-formedness of the
+        // extension set.
+        let leaf_san = &chain[0]
+            .certificate()
+            .extensions()
+            .unwrap()
+            .get_extension(&SUBJECT_ALTERNATIVE_NAME_OID)
+            .unwrap();
+
```

Comment:
You need to handle what if the SAN has GNs that are of unsupported types here.

Reply:
I did this by propagating the error out of `x509::parse_general_names` -- is that what you were thinking? I think the GN parsing here can't directly fail, since the Rust-only `GeneralName` type has variants for all possible GN forms, whereas the Python side that's produced with `x509::parse_general_names` is what doesn't.

Reply:
Sorry, when I said unsupported types I specifically meant: anything besides DNSName + IPAddress (the thing we know how to do nameconstraint checks for).

If you have a SAN with both an EDIPartyName and an IPAddress, either the valid names should be [IP] only, or it should be an error -- not sure which.

Reply:
Makes sense, thank you for clarifying!

One question: when there are no name constraints but the SAN contains GNs other than DNS and IPs, should we return those? My mental model for how client certs typically convey subjects is a little murky, but I can imagine that both email addresses and X.500 names are somewhat common (and wouldn't necessarily have NCs).

Reply:
Seems like a big compatibility hazard to do that: Just adding _any_ NC to any CA would break that.

Seems like really we should implement NC checking for emails at a minimum, in a seperate PR, before merging this PR.

Reply:
Okay, we now handle NCs for unsupported types more correctly (https://github.com/pyca/cryptography/pull/10570) and surface `UnsupportedGeneralNameType` if any of the leaf's SANs don't have a corresponding Python type.